java/java.lsp.server/vscode/src/explorer.ts

```diff
@@ -0,0 +1,253 @@
+import * as vscode from 'vscode';
+import {  LanguageClient } from 'vscode-languageclient/node';
+import { NodeChangedParams, NodeInfoNotification, NodeInfoRequest } from './protocol';
+
+class TreeViewService {  
+  private client : LanguageClient;
+  private trees : Map<string, vscode.TreeView<Visualizer>> = new Map();
+  private images : Map<number, vscode.Uri> = new Map();
+  private providers : Map<number, VisualizerProvider> = new Map();
+  constructor (c : LanguageClient) {
+    this.client = c;
+  }
+
+  getClient() : LanguageClient {
+    return this.client;
+  }
+
+  async createView(id : string, options? : Partial<vscode.TreeViewOptions<any>>) : Promise<vscode.TreeView<Visualizer>> {
+    let tv : vscode.TreeView<Visualizer> | undefined  = this.trees.get(id);
+    if (tv) {
+      return tv;
+    }
+    const res = await createViewProvider(id);
+    this.providers.set(res.getRoot().data.id, res);
+    let opts : vscode.TreeViewOptions<Visualizer> = {
+      treeDataProvider : res,
+      canSelectMany: true,
+      showCollapseAll: true,
+    }
+    
+    if (options?.canSelectMany !== undefined) {
+      opts.canSelectMany = options.canSelectMany;
+    }
+    if (options?.showCollapseAll !== undefined) {
+      opts.showCollapseAll = options.showCollapseAll;
+    }
+    let view = vscode.window.createTreeView(id, opts);
+    // this will replace the handler over and over, but never mind
+    this.client.onNotification(NodeInfoNotification.type, params => this.nodeChanged(params));
+    return view;
+  }
+
+  private nodeChanged(params : NodeChangedParams) : void {
+    let p : VisualizerProvider | undefined = this.providers.get(params.rootId);
+    if (p) {
+      p.refresh(params);
+    }
+  }
+
+  imageUri(nodeData : NodeInfoRequest.Data) : vscode.Uri | undefined {
+    if (nodeData.iconUri) {
+      const uri : vscode.Uri = vscode.Uri.parse(nodeData.iconUri)
+      this.images.set(nodeData.iconIndex, uri);
+      return uri;
+    } else {
+      return this.images.get(nodeData.iconIndex);
+    }
+  }
+}
+
+let treeService : Promise<TreeViewService>;
+let promisedTreeService : (ts : TreeViewService) => void;
```

Comment:
Please eliminate all static state from this `.ts` file and other `.ts` files. The only static state shall be in `explorer.ts`, currently:
```ts
let client: Promise<LanguageClient>;
let testAdapter: NbTestAdapter | undefined;
let nbProcess : ChildProcess | null = null;
let debugPort: number = -1;
let consoleLog: boolean = !!process.env['ENABLE_CONSOLE_LOG'];
```
Make the `TreeViewService` a field in `LanguageClient`, if possible.

Reply:
Done (by you :-)) in 60a33320